Introducing a new option: maxOccupationTime

abozaralizadeh opened the pull request:
**maxWaitTime** is another form of timeout that covers some use-cases that **timeout** option might not be useful. 
In case of **timeout**, Async-lock would throw an error, and then the queue will lose all of its waiting items.
**maxWaitTime** is a timeout for every single item and if the function exceeds the time, we release the lock and let the next task acquire the lock without canceling the unfinished task.

In collaboration with @ThePiz, we suggest this new option as we are using Async-lock in our project & this feature is essential for our use case.

rogierschouten commented:
Hi @abozaralizadeh 

What is your use case exactly?  Releasing locks without cancelling the task sounds like it makes everything you do from that point onward would be incorrect/unpredictable?   What guarantees would the user still have after that point?

There is one likely bug in your code, namely that you never cancel the timer. I think it should be canceled wherever the lock is released for another reason than your wait timeout. 

If you want this integrated, please also create unit tests (preferably better ones than the existing ones, e.g. use sinon)  and update README.md to explain your new feature. 

If that would be too much work for you, you might consider just forking the repository and making it work for your project.  You can directly reference repositories in your package.json, so you don't actually need it to be published.

abozaralizadeh commented:
Hi @rogierschouten

Thank you for your comment.

Locks can be used not only for protecting critical sections but also in other use-cases such as avoiding race conditions or guaranteeing arbitrary ordering. In our case, the Async-lock helped us to implement a fairly ordered message delivery system using unordered Ack messages. the lock would wait for the Ack message to unlock the next message, but in case of losing the Ack, we implemented this new option to unlock the queue anyway after exceeding the max occupation time.

We are already using our fork in our project but I was sure that this option might be useful also for others so I decided to suggest the current PR.

After reading your comment I decided to generalize the option and raise an error to give the handling part to the programmer to decide. it might be 'not canceling the task' and just releasing the lock as in our use-case or canceling it and do whatever is needed to be done (although some jobs can’t be simply undone/canceled)

As you asked I also added related unit test and modified the README.md file.

Thank you

rogierschouten commented:
That is starting to look really good!  I still think that you need to cancel the timer in more places?  Now, when the lock is released, the timer still runs.  Maybe in the `done` function?  
Please tell me if I'm wrong, I am not really into this code.

abozaralizadeh commented:
Thanks! actually, I believe that right now it should be just fine. every time the lock is released naturally before the timer, 

**1.** If there is another task in the queue, it will clear the timer here:
```javascript
			if (occupationTimer) {
				clearTimeout(occupationTimer);
			}
```
and set the new timer quickly afterward.
```javascript
			occupationTimer = setTimeout(function () {...});
```
**2.** If the task is the last one in the queue, the queue will be deleted, so when the timer due time arrives and runs it's function, the first condition `if (!!self.queues[key])` will be false, so basically nothing happens:
```javascript
	occupationTimer = setTimeout(function () {
		if (!!self.queues[key]) {
			done(false, new Error('Maximum occupation time is exceeded'));
		}
	}, maxOccupationTime);
```
Finally, in case of error due to '_Maximum occupation time is exceeded_', it's the programmer's job to handle the error and release the lock as described in the previous comment.

rogierschouten commented:
`f the task is the last one in the queue, the queue will be deleted, so when the timer due time arrives and runs it's function, the first condition if (!!self.queues[key]) will be false, so basically nothing happens:`

Yeah that is what I meant, even though it is guarded, it is a timer that runs unnecessarily.  Let me know if you still want to fix it or not, I will merge the PR otherwise.

abozaralizadeh commented:
I see, I just moved the reset part to the very beginning of `done` and that should be more accurate as it will reset the timer every time after release/error. I think this is better than the previous commit as it solves both unnecessary timer in the last item and also reset the timer in any other case of error. Thanks

rogierschouten commented:
Thank you @abozaralizadeh and @ThePiz 

Published as v1.3.0.    I forgot to add both of you as contributors so I did that in the next commit after the release.
